internal/rollout/device_selection/batch_sequence_selector.go

```diff
@@ -0,0 +1,520 @@
+}
+
+func (b *batchSequenceSelector) Reset(ctx context.Context) error {
+	b.log.Infof("%v/%s:In Reset", b.orgId, b.fleetName)
+	if err := b.UnmarkRolloutSelection(ctx); err != nil {
+		return err
+	}
+	if err := b.resetApprovalAndSuccessPercentage(ctx); err != nil {
+		return err
+	}
+	return b.setCurrentBatch(ctx, -1)
+}
+
+type batchSelection struct {
+	batch               *api.Batch
+	store               store.Store
+	orgId               uuid.UUID
+	fleetName           string
+	templateVersionName string
+	fleet               *api.Fleet
+	updateTimeout       time.Duration
+	log                 logrus.FieldLogger
+}
+
+func (b *batchSelection) getAnnotation(annotation string) (string, bool) {
+	annotations := lo.FromPtr(b.fleet.Metadata.Annotations)
+	if annotations == nil {
+		return "", false
+	}
+	v, exists := annotations[annotation]
+	return v, exists
+}
+
+func (b *batchSelection) IsApproved() bool {
+	approvedStr, exists := b.getAnnotation(api.FleetAnnotationRolloutApproved)
+	if !exists {
+		return false
+	}
+	return approvedStr == "true"
+}
+
+func (b *batchSelection) IsRolledOut(ctx context.Context) (bool, error) {
+	// A rolled out device is one that its template version is the same as the expected template version
+	// A rolled out batch is a batch that for every of its devices there is no device that its template
+	// version is not equal to the expected template version
+	newQuerySelectorParts().withOwner(b.fleetName).withSelectedForRollout().withoutRolledOut(b.templateVersionName).listParams()
+	count, err := b.store.Device().Count(ctx, b.orgId, newQuerySelectorParts().
+		withOwner(b.fleetName).
+		withSelectedForRollout().
+		withoutRolledOut(b.templateVersionName).
+		listParams())
+	if err != nil {
+		return false, err
+	}
+	return count == 0, nil
+}
+
+func (b *batchSelection) getSuccessThreshold() (int, error) {
+	var (
+		ret int
+		err error
+	)
+	successThreshold := b.fleet.Spec.RolloutPolicy.SuccessThreshold
+	if b.batch != nil && b.batch.SuccessThreshold != nil {
+		successThreshold = b.batch.SuccessThreshold
+	}
+	if successThreshold != nil {
+		ret, err = api.PercentageAsInt(*successThreshold)
+		if err != nil {
+			return 0, err
+		}
+	} else {
+		ret = DefaultSuccessThreshold
+	}
+	return ret, nil
+}
+
+func (b *batchSelection) getLastSuccessPercentage() (int, bool, error) {
+	percentageStr, exists := b.getAnnotation(api.FleetAnnotationLastBatchSuccessPercentage)
+	if !exists {
+		return 0, false, nil
+	}
+	ret, err := strconv.ParseInt(percentageStr, 10, 64)
+	if err != nil {
+		return 0, false, err
+	}
+	return int(ret), true, nil
+}
+
+// A batch may be approved atotmatically only if its approval method is "automatic" and the
+// success percentage of the previous batch is greater or equal to the success threshold
+func (b *batchSelection) MayApproveAutomatically() (bool, error) {
+	approvalMethod, exists := b.getAnnotation(api.FleetAnnotationRolloutApprovalMethod)
+	if !exists || approvalMethod != "automatic" {
```

Comment:
if this doesn't exist I would expect approvalMethod to always be empty.  if its not might be something to look at.

Reply:
OK